### PR TITLE
Add back missing functionality needed for cvent

### PR DESCRIPTION
**Pull request by brentryan**
These are the changes/fixes we need to support our use cases at Cvent.  Some of the tests/code used to validate this are discussed here https://github.com/jonpeterson/jackson-module-model-versioning/pull/4

There are really a few scenarios I'd like out of this (I'm open to other options):
1.  I should be able to do this and it work properly:

```
    @JsonSerializeToVersion(defaultToSource = true)
    private String serializeToVersion;
```
1.  I should be able to deserialize JSON with 0 no modelVersion in it and serialize back out to JSON with no modelVersion defined
2.  I should be able to deserialize JSON with modelVersion of 1 while my concrete class is on version 3 and when I use @JsonSerializeToVersion(defaultToSource = true) it should serialize back out to modelVersion of 1 (not 3)

These changes fix all 3 cases above.  Hopefully this is helpful.

### CODE REVIEW

**Comment by jonpeterson** [comment]
I believe this PR is overkill.  I have written some test code to prove that it is not needed.

``` java
public class TestCase {
    private static final ObjectMapper objectMapper = new ObjectMapper().registerModule(new VersioningModule());

    public static void main(String[] args) throws IOException {
        printCase("{\"a\": 1}");
        printCase("{\"a\": 2, \"modelVersion\": \"1\"}");
        printCase("{\"b\": 3, \"modelVersion\": \"2\"}");
    }

    private static void printCase(String serialized) throws IOException {
        TestModel deserialized = objectMapper.readValue(serialized, TestModel.class);
        String reserialized = objectMapper.writeValueAsString(deserialized);

        System.out.println();
        System.out.println("serialized:   " + serialized);
        System.out.println("reserialized: " + reserialized);
    }


    @JsonVersionedModel(
        currentVersion = "2",
        defaultDeserializeToVersion = "0",
        toCurrentConverterClass = TestModelToCurrentConverter.class,
        toPastConverterClass = TestModelToPastConverter.class
    )
    public static class TestModel {

        public int b;

        @JsonProperty
        @JsonSerializeToVersion(defaultToSource = true)
        private String serializeToVersion;
    }

    public static class TestModelToCurrentConverter implements VersionedModelConverter {

        @Override
        public ObjectNode convert(ObjectNode modelData, String modelVersion, String targetModelVersion, JsonNodeFactory nodeFactory) {
            int modelVersionNum = Integer.parseInt(modelVersion);

            if(modelVersionNum < 2)
                modelData.put("b", modelData.remove("a").intValue());

            return modelData;
        }
    }

    public static class TestModelToPastConverter implements VersionedModelConverter {

        @Override
        public ObjectNode convert(ObjectNode modelData, String modelVersion, String targetModelVersion, JsonNodeFactory nodeFactory) {
            int targetModelVersionNum = Integer.parseInt(targetModelVersion);

            if(targetModelVersionNum < 2)
                modelData.put("a", modelData.remove("b").intValue());
            if(targetModelVersionNum == 0)
                modelData.remove("modelVersion");

            return modelData;
        }
    }
}
```

When executing with 1.2.0 (what I released last night), the output is:

```
serialized:   {"a": 1}
reserialized: {"a":1,"modelVersion":"0"}

serialized:   {"a": 2, "modelVersion": "1"}
reserialized: {"a":2,"modelVersion":"1"}

serialized:   {"b": 3, "modelVersion": "2"}
reserialized: {"b":3,"modelVersion":"2"}
```

After a single-line change to `VersionedModuleSerializer`, the output is:

```
serialized:   {"a": 1}
reserialized: {"a":1}

serialized:   {"a": 2, "modelVersion": "1"}
reserialized: {"modelVersion":"1","a":2}

serialized:   {"b": 3, "modelVersion": "2"}
reserialized: {"b":3,"modelVersion":"2"}
```

The important thing to note is the `@JsonProperty` on the `private String serializeToVersion;`.

I will commit the single line fix to the cvent branch for you to test.

Please confirm that this meets your project requirements.

**Comment by brentryan** [comment]
So I figured out the issue with my test.  When you have these on your model

```
    public String getSerializeToVersion() {
        return serializeToVersion;
    }

    public void setSerializeToVersion(String serializeToVersion) {
        this.serializeToVersion = serializeToVersion;
    }
```

Then this causes it not to work as expected.  The only way I could get it to work is to change to this 

```
     private String serializeToVersion;

    @JsonSerializeToVersion(defaultToSource = true)
    public String getSerializeToVersion() {
        return serializeToVersion;
    }

    public void setSerializeToVersion(String serializeToVersion) {
        this.serializeToVersion = serializeToVersion;
    }
```

While this works, I think it's more headache for the consumers of this to have to worry about putting the annotation in the right place.  I'd also rather not have to put @JsonProperty on the field.  My preference is for this to work:

```
    @JsonSerializeToVersion(defaultToSource = true)
    private String serializeToVersion;

    public String getSerializeToVersion() {
        return serializeToVersion;
    }

    public void setSerializeToVersion(String serializeToVersion) {
        this.serializeToVersion = serializeToVersion;
    }
```

Am I crazy?

The only other thing I'm not crazy about it having to create a converter just to remove "modelVersion" even if I do not want to provide any converter by default.  If I have a model with no versioning support and then later decide to add versioning support (modelVersion=2) but I don't want to make a converter because I'm backwards compatible then it seems a bit overkill to have to require the developer to create a past converter just to remove the modelVersion property.  Yes it works, just not ideal IMO.

**Comment by jonpeterson** [comment]
Ok, I see the issue now with the field and getters annotations not merging.  I know that Jackson is designed to be able to inherit annotations from both fields and getters/setters so aren't supposed to have to care about writing code paths for both, but for some reason it's not doing it until I call `((POJOPropertyBuilder)definition).mergeAnnotations(true);`.  Adding that to the `VersionedModelUtils.getSerializeToVersionProperty` solves your first issue.

As for the second part: I see now and agree. I have cherry picked in the `if (!targetVersion.equals(jsonVersionedModel.defaultDeserializeToVersion()))` part.  I am maintaining compatibility with Java 1.6 (as per Jackson Core), so I left out the try-with-resources bit.

Thanks for the detailed explanation of the issues.  I'll get this pushed to the branch.  Hopefully it will address both your issues.

**Comment by brentryan** [comment]
Awesome. Thanks man!

**Comment by jonpeterson** [comment]
Great, I'll release 1.2.1 this evening then.

**Comment by jonpeterson** [comment]
I released 1.2.1 with a slight change.  After thinking about it more, I'm not sure if all use cases lend themselves to having `defaultDeserializeToVersion == hide version`, so I made another property on the `@JsonVersionedModel` annotation called `versionToSuppressPropertySerialization`.  I was having a hard time coming up with a concise variable name, lol.  Anyway, the idea is that when that property is set and when it is equal to the target serialized version, the modelVersion property isn't serialized.

Basically, for your use case, just add `versionToSuppressPropertySerialization` to whatever `defaultDeserializeToVersion` is set to.

**Comment by brentryan** [comment]
That works! Thanks for putting in the effort.
